test_gpt2.cu

```diff
@@ -8,33 +8,43 @@ int check_tensor(float *a, float *b, int n, const char* label, float threshold=1
     int ok = 1;
     float max_diff = 0.0f;
     float max_rel_error = 0.0f;
+    float max_to_threshold = 0.f;
     float max_a = 0.0f;
     float max_b = 0.0f;
-    printf("%s\n", label);
+    float epsilon = 0.079;      // BF16 epsilon value
+    printf("%8s: ", label);
     for (int i = 0; i < n; i++) {
+        float t_eff = threshold + fabs(b[i]) * epsilon;
         float diff = fabsf(a[i] - b[i]);
+        max_to_threshold = max(max_to_threshold, diff / t_eff);
         if (diff > max_diff) {
             max_diff = diff;
             float denom = fabsf(b[i]);
             max_rel_error = (denom == 0.0f) ? 0.0f : diff / denom;
             max_a = a[i];
             max_b = b[i];
         }
-        if (diff <= threshold) {
-            if (i < print_upto) { printf("OK "); }
-        } else {
-            if (i < print_upto) { printf("NOT OK "); }
+        if (diff > t_eff) {
             ok = 0;
         }
-        if (i < print_upto) { printf("%f %f\n", a[i], b[i]); }
     }
     // print the final result
     if (ok) {
-        printf("TENSOR OK, max diff: %e, with rel error: %e (calculated=%f, ref=%f)\n",
-                max_diff, max_rel_error, max_a, max_b);
+        printf("TENSOR OK, max diff: %.3e, with rel error: %.3e (calculated=%10f, ref=%10f), %.2f%% of maximum error\n",
+                max_diff, max_rel_error, max_a, max_b, max_to_threshold*100);
     } else {
-        printf("TENSOR NOT OK, max diff: %e, with rel error: %e (calculated=%f, ref=%f)\n",
-                max_diff, max_rel_error, max_a, max_b);
+        printf("TENSOR NOT OK, max diff: %.3e, with rel error: %.3e (calculated=%10f, ref=%10f), %.2f%% of maximum error\n",
+                max_diff, max_rel_error, max_a, max_b, max_to_threshold*100);
+    }
+
+    if(ok == 0) {
+        for (int i = 0; i < print_upto; i++) {
+            float t_eff = threshold + fabs(b[i]) * epsilon;
+            float diff = fabsf(a[i] - b[i]);
+            printf(diff <= threshold ? "OK " :  "NOT OK ");
```

Comment:
i think this condition here is a bug, it should match the one up above, i'll fix.

Reply:
yes